libibverbs/memory.c

```diff
@@ -57,68 +57,99 @@ struct ibv_mem_node {
 	int			refcnt;
 };
 
+#define MAX_NUM_PAGE_SIZES 32
+
 static struct ibv_mem_node *mm_root;
 static pthread_mutex_t mm_mutex = PTHREAD_MUTEX_INITIALIZER;
-static int page_size;
 static int huge_page_enabled;
 static int too_late;
+static int num_page_sizes;
+static unsigned long page_sizes[MAX_NUM_PAGE_SIZES];
 
-static unsigned long smaps_page_size(FILE *file)
-{
-	int n;
-	unsigned long size = page_size;
-	char buf[1024];
+enum {
+	IBV_SC_PAGE_SIZE_IDX = 0,
+};
 
-	while (fgets(buf, sizeof(buf), file) != NULL) {
-		if (!strstr(buf, "KernelPageSize:"))
+static unsigned long get_page_size(void *base)
+{
+	unsigned long ret = page_sizes[IBV_SC_PAGE_SIZE_IDX];
+	unsigned long base_aligned;
+	unsigned long pgsz;
+	int err;
+	int i;
+
+	for (i = 0; i < num_page_sizes; i++) {
+		pgsz = page_sizes[i];
+		base_aligned = (uintptr_t)base & ~(pgsz - 1);
+		/* DOFORK is required as the kernel caches the given flags,
+		 * so calling madvise with MADV_DONTFORK twice on the same vma
+		 * returns success without checking the address and length.
+		 */
+		while ((err = madvise((void *)base_aligned, pgsz,
```

Comment:
This doesn't look like it should work at all. madvise will split VMAs, so if you call this with 4k it will happily split the huge page VMA into sub VMAs and go ahead.

The entire point of this nonsense code is to prevent splitting the VMA, which can only be done by learning the actual extents of the VMA and acting accordingly.

If it is too slow then don't use the broken forking mode. It doesn't actually work reliably anyhow.

Reply:
> This doesn't look like it should work at all. madvise will split VMAs, so if you call this with 4k it will happily split the huge page VMA into sub VMAs and go ahead.

I thought so as well, apparently it doesn't. It returns an EINVAL error.

> The entire point of this nonsense code is to prevent splitting the VMA, which can only be done by learning the actual extents of the VMA and acting accordingly.
> 
> If it is too slow then don't use the broken forking mode. It doesn't actually work reliably anyhow.

Didn't know there are forking modes? And that one of them is broken?

Reply:
How does the kernel flow generate EINVAL? I don't see it, I see VMA splitting.

By forking mode I mean don't use fork, so don't use this madvise marking

Reply:
I can't point the exact flow where this happens in the kernel, but running this sample app:
```
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>
#include <stdint.h>
#include <errno.h>

#define PAGE_SIZE (4096)
#define HUGEPAGE_SIZE (1024 * 1024 * 2)

int main()
{
	void *buf;
	void *buf_aligned;
	int err;
	int pgsz;
	int sizes[2] = { PAGE_SIZE, HUGEPAGE_SIZE };
	int i;

	buf = mmap(NULL, 64, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB,
		   -1, 0);
	if (!buf) {
		printf("Couldn't allocate memory\n");
		return -1;
	}

	printf("buf %p\n", buf);

	for (i = 0; i < 2; i++) {
		pgsz = sizes[i];
		buf_aligned = (void *)((uintptr_t)buf & ~(pgsz - 1));
		err = madvise(buf_aligned, pgsz, MADV_DONTFORK);
		printf("madvise pgsz %d err %d errno %d\n", pgsz, err, err ? errno : 0);
	}

	munmap(buf, 64);
	return 0;
}
```

Outputs:
```
buf 0x7f0b0f800000
madvise pgsz 4096 err -1 errno 22
madvise pgsz 2097152 err 0 errno 0
```

Do you suggest that all libibverbs apps don't use fork?

Reply:
We can't rely on some kernel behavior like this without understanding if it is guarenteed or not, since I see code that is doing VMA splitting in madvise you need to figure out why it is isn't activting and if that is part of the API contract, then document all of this.

And yes, verbs apps using fork is very problematic, it should be avoided.

Reply:
Thanks, I agree we shouldn't rely on this behavior without a better understanding.
Let's drop this patch for now.